### ISSUE TITLE
Add Git integration for themes

### ISSUE THREAD

**Issue by madhusudhand** [WordPress/create-block-theme#435]
Add ability to connect a git repository with a WordPress site to import or commit theme changes.

### In wp-admin: 

WP-admin interface presents a "Create Block Theme Git Utilities" menu.
This shows allows to connect/disconnect a git repository and shows details such as

* Option to connect a repository if it is not already connected
* Option to disconnect existing repository
* List of connected/installed themes between WP-env and Git repo
* And may be status of their sync status
* List of available branches in the connected repo
* List of available themes (after selecting a branch)
* Allows to select single theme and install (many be multiple)?

### In the site editor

Under the 'Create Block Theme' (tool icon) menu another option 'Git Utilities' is added.

It has the following options/purpose:

* Shows if a repository is connected or not
* If a repository is not connected, it gives a link that takes user back to above wp-admin menu (It should allow to connect repo only from admin-menu)
* Connected repository URL is shown (read only)
* User is presented with list of available branches and can switch a branch
* Or create a new branch

### Connect

**Where?** From the wp-admin interface

User is presented an input where they can enter the url to a git repository (Https url). The option to include a token is available but a URL is all that is required. (Without a key only 'pulling' actions occur on public repos). Key is required if repo is private or need to push changes.

Once they enter the URL:
(IF it is determined that multiple themes are present in the repository) The user is presented with a list of themes that can be installed. Once they select one:
The user is given confirmation to install the (selected/only) theme which is then "git installed" and activated. (permission to install a theme is checked)

> Question:  
After connecting repo, if only one theme is selected and installed, are we going to consider one theme as connected to pull or push? Or is this selection of themes is just for installation?  
I think only selected themes must stay connected even if repository has many other themes. and changes of those themes can be pulled or pushed. Thoughts?

###  Disconnect

An option "Sever Connection to Git Repository" is presented in the wp-admin page to disconnect the repository. By doing so, it simply removes the clone of the git repository and doesn't effect already installed/pulled themes or changes.

It must be presented to user as a prompt before disconnecting.

### Branch switching

Branch can be switched either from wp-admin or site-editor.
List of available branches in the repo are freshly fetched and presented to the user.

User can choose a branch and only themes under it are shown for selection.

User can also choose to create a new branch. New branch is always created from remote HEAD (trunk/master/main etc.)

## Pull / Push scenarios to consider

| WP site | Git repo | In Sync? | Pull | Push |
| --- | --- | --- | - | - |
| theme ❌ | theme ✅ | - | ✅ | - |
| theme ✅ | theme ❌ | - | - | ✅ |
| theme ✅ | theme ✅ | In Sync | ✅ | ✅ |
| theme ✅ + changes 👨‍💻 | theme ✅ | No | discard local changes | ✅ |
| theme ✅ | theme ✅ + changes 👨‍💻 | No | ✅ | Pull to sync |
| theme ✅ + changes 👨‍💻 | theme ✅ + changes 👨‍💻 | No | discard local changes | ❌ |

### Discarding changes:

In cases where local changes has to be discarded, user will be presented with two options. Either to delete them locally or create a separate branch to save those changes (this new branch is automatically created with format such as current-branch-timestamp)

### Handling Not in sync changes

When changes are not in sync (any of the above scenarios), user has to bring them to sync either by Discarding, pulling or pushing.
Only after the branches are brought to sync, user theme changes are identified and the diff is shown.

### Push

User changes are pushed to remote repo immediately after performing a commit. It should always be a clean fast-forward push. If there is additional commit in the remote branch, it should have been identified as not-in-sync in the before allowing user to commit.

### Pull

User can pull updates for the selected branch to bring them to sync. After a pull user is asked to replace the theme updates.

> Question: which theme updates? Do we present all themes again or only the current theme updates?

### Diff and Commit

A diff of changes is shown and commit is allowed only once the branches are in sync (commitable state).
There is not explicit push choice for the user. A commit in the UI means, commit + push to remote.

This eliminates the scenario where local branch has some changes that aren't there in remote and confusion of not seeing changes  in the PR for non-tech user)

### Creating a Pull request

Out of scope. User need to create the Pull Request using the web interface of git hosting platform (such as github)

----------

## Tasks: [Draft version]

Push theme changes from WP site to Git repository:

- [ ] Link git repositories with the WP environment from the editor - ⏳ @madhusudhand 
- [ ] Create/fetch a remote branch
- [ ] Show theme changes and allow commit (&push) to remote - ⏳ @madhusudhand 
- [ ] Allow a repository to be linked with a single theme or all themes of the WP site - ⏳

Pull theme changes from Git repository to WP site:

- [ ] Link a new theme repository (single or many themes) and allow user to fetch/install set of themes
- [ ] Fetch changes from an upstream branch into WP site

Both ways:

- [ ] Identify conflicts and show appropriate errors to the user (conflict resolution isn't in the scope of this integration)
- [ ] Disconnect a repository with WP site - ⏳

**Comment by pbking** [comment]
I made a slightly different list from the above.  I tried to capture the workflows of somebody using this too.  Here are the few I came up with:

### Install a theme from Github (trunk).
From the wp-admin interface a user selects the "Create Block Theme Git Utilities" menu.
A user is presented an input where they can enter the url to a git repository.  The option to include a token is available but a URL is all that is required.  (Without a key only 'pulling' actions occur.)
Once they enter the URL:
 (IF it is determined that multiple themes are present in the repository) The user is presented with a list of themes that can be installed.  Once they select one:
 The user is given confirmation to install the (selected/only) theme which is then "git installed" and activated.

### Switch to another branch of a theme's repo (For example when a user needs to test and validate a bug fix in a branch)
From the Site Editor, in the 'Create Block Theme' menu a user selects the 'Git Utilities' menu item.
There they see that the active theme is "git installed".
(If they have un-synced changes they cannot continue until the changes have either been PUSHED or DROPPED.)
They are given a list of branches available in the repository. (bonus points for filtering out any branches in which the current theme isn't available)
A user chooses a branch which is then synced to that environment.

### Install a theme only available in a branch (For example when a user needs to test and validate a new theme)
These are the same steps as 'Install a Theme from Github" except when the user is presented with the list of themes available they change the `/trunk` branch to any other branch.  Then the themes in that branch are presented as options to select from.

### Pull updates from the remote branch to local.
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
There they see that the installed theme is "out of date" from the remote and are given the option to "sync" it.  
(If they have un-synced changes they cannot continue until the changes have either been PUSHED or DROPPED.)
Clicking it pulls the changes.

### Drop Local Changes
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
They see that they have made changes and wish to 'DROP' those changes and return their theme to the state of the last committed change.
They click the 'DROP CHANGES' button and, after confirming that that's what they intended, the changes are dropped.

### Sever Git Connection
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
From there they click on a 'settings' menu item (gear icon?)
They click the 'Sever Connection to Git Repository' button and, after confirming that that is what they want to do the theme is removed from the plugin repository, the theme remains installed, and the 'Git Utilities' menu shows in the "disconnected" state.

### Install a theme from a git repository that is already installed locally
The user steps are the same as 'Install a theme from a Git Repository' except that, if the theme is ALREADY installed a user must choose between DROPPING any changes (to use the current branch) and CREATING A NEW BRANCH.  (Note that if there are no changes nothing special should need to happen)

### Add/change a token to the git configuration of a theme.
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
From there they click on a 'settings' menu item (gear icon?)
They are shown a read-only URL for the connection to the git repository.
Additionally there is a token field that shows as filled (obscured) if a user has already entered a token, or empty if they have not.  
A user can enter a token here and click 'update' to store that new token.

### Connect a local, active theme to a git repository in which it already exists
(This is functionally the same as 'Install a theme from a git repository that is already installed locally) but achieved through different steps)
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
They see that the theme is in a "disconnected" state and they have the option to enter the URL of a git repository the theme is stored in.
They are also given the option to provide their token at this point (but it is not required).
A user is able to set a folder the theme is expected to be found in.  Perhaps this defaults to `/` but `${THEME_SLUG}` is provided as a selectable option?  Or a full path such as `wp-content/themes/pickle`.
Once the theme has been checked out and a theme (any theme) is found to be at that location, if there are any differences, the user is given the option to DROP local changes or create a new (local, disconnected) branch.

### Connect a local, active theme to a git repository in which it does NOT exist
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
They see that the theme is in a "disconnected" state and they have the option to enter the URL of a git repository the theme will be stored in.
They are also given the option to provide their token at this point (but it is not required).
A user is able to set a folder the theme is expected to reside in.  Perhaps this defaults to `/` but `${THEME_SLUG}` is provided as a selectable option?  Or a full path such as `wp-content/themes/pickle`.
When it is found that no theme exists at that location a new (local, disconnected) branch is created.

--- The following can only be done with a valid token ---

### Push local changes to an existing theme to a NEW BRANCH and create a PULL REQUEST
When a theme is checked out in a branch that is local & disconnected,
From the Site Editor, in the 'Create Block Theme' menu a user select the 'Git Utilities' menu item.
They see that the theme is in a "connected" but "unsyncable" state.
A user is shown all of the changed files and how they were changed (added, deleted, etc)
The user is present the option to "PUSH" their changes, prompted for a reasonable branch name, comments regarding the change, and details that will go in the Pull Request.

**Comment by madhusudhand** [comment]
Thanks @pbking for adding thoughts. I have summarized your thoughts slightly differently along with thoughts I have in the issue description.

> Switch to another branch of a theme's repo: They are given a list of branches available in the repository. (bonus points for filtering out any branches in which the current theme isn't available)

I guess the filtering isn't required. User should be able to switch to any branch without restrictions (because witching doesn't remove or add themes), and based on whether the themes exists (by checking if they are in sync) user is allowed either pull (update existing/ install new) or push changes to the branch.

> create a PULL REQUEST

I think for the first iteration, we can just take care of pushing to a remote branch. Let user create a Pull request themselves.

**Comment by pbking** [comment]
No long plans for CBT to interact with a specific version control.  Usage with playground is now the closest thing I think.
